### PR TITLE
Add generic open datasources so that each project don't need their own

### PR DESCRIPTION
**Pull request by sdetilly**
## Description

Every project I worked on had these generic classes. It is simpler if the open, generic classes that we want to overload come from the lib instead.

## Motivation and Context

I wanted to stop copying the boilerplate from projet to project

## How Has This Been Tested?

I took these classes from an existing project

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: trikot-datasources/datasources-flow/src/jsMain/kotlin/com/mirego/trikot/datasources/flow/filesystem/NativeFileSystem.js.kt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package com.mirego.trikot.datasources.flow.filesystem
+
+import okio.FileSystem
+import okio.NodeJsFileSystem
+
+actual object NativeFileSystem {
```
Comment (npresseault) [inline trikot-datasources/datasources-flow/src/jsMain/kotlin/com/mirego/trikot/datasources/flow/filesystem/NativeFileSystem.js.kt:6]: Does this need to be public?

File: trikot-datasources/datasources-flow/src/commonMain/kotlin/com/mirego/trikot/datasources/flow/generic/GenericDataSource.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+import com.mirego.trikot.datasources.flow.BaseExpiringExecutableFlowDataSource
+import com.mirego.trikot.datasources.flow.DataSourceUtils
+import com.mirego.trikot.datasources.flow.ExpiringFlowDataSourceRequest
+import com.mirego.trikot.datasources.flow.FlowDataSourceRequest
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.json.Json
+import kotlin.time.Duration.Companion.minutes
+
+open class GenericDataSource<T>(
```
Comment (npresseault) [inline trikot-datasources/datasources-flow/src/commonMain/kotlin/com/mirego/trikot/datasources/flow/generic/GenericDataSource.kt:11]: Why this? It already exists in datasources-flow

```
data class ExpiringExecutableFlowDataSourceRequest<T>(
    override val cacheableId: String,
    override val expiredInMilliseconds: Long,
    override val requestType: FlowDataSourceRequest.Type = FlowDataSourceRequest.Type.USE_CACHE,
    val block: suspend () -> T
) : ExpiringFlowDataSourceRequest

class ExpiringExecutableFlowDataSource<T>(
    cacheDataSource: FlowDataSource<ExpiringExecutableFlowDataSourceRequest<T>, FlowDataSourceExpiringValue<T>>? = null,
    coroutineContext: CoroutineContext = EmptyCoroutineContext
) : BaseExpiringExecutableFlowDataSource<ExpiringExecutableFlowDataSourceRequest<T>, T>(cacheDataSource, coroutineContext) {

    override suspend fun internalRead(request: ExpiringExecutableFlowDataSourceRequest<T>) =
        FlowDataSourceExpiringValue(request.block(), Date.now.epoch + request.expiredInMilliseconds)
}
```

File: trikot-datasources/datasources-flow/src/commonMain/kotlin/com/mirego/trikot/datasources/flow/generic/GenericDataSource.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    json: Json,
+    dataSerializer: KSerializer<T>,
+    diskCachePath: String?
+) : BaseExpiringExecutableFlowDataSource<GenericDataSourceRequest<T>, T>(
+    diskCachePath?.let { GenericDiskDataSource(json, dataSerializer, diskCachePath) }
+) {
+    override suspend fun internalRead(request: GenericDataSourceRequest<T>) = DataSourceUtils.buildExpiringValue(request.block(), request)
+}
+
+data class GenericDataSourceRequest<T>(
```
Comment (npresseault) [inline trikot-datasources/datasources-flow/src/commonMain/kotlin/com/mirego/trikot/datasources/flow/generic/GenericDataSource.kt:21]: That thing also exist and is also called `ExpiringExecutableFlowDataSourceRequest`

File: trikot-datasources/datasources-flow/src/commonMain/kotlin/com/mirego/trikot/datasources/flow/generic/GenericDiskDataSource.kt [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+import com.mirego.trikot.datasources.flow.filesystem.NativeFileSystem
+import kotlinx.coroutines.withContext
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.SerializationException
+import kotlinx.serialization.json.Json
+import okio.FileNotFoundException
+import okio.Path
+import okio.Path.Companion.toPath
+
+class GenericDiskDataSource<R : ExpiringFlowDataSourceRequest, T>(
```
Comment (npresseault) [inline trikot-datasources/datasources-flow/src/commonMain/kotlin/com/mirego/trikot/datasources/flow/generic/GenericDiskDataSource.kt:16]: Why generic? I'd go with a shorter term like
FileSystemDataSource or something like that.

File: trikot-datasources/datasources-flow/src/commonMain/kotlin/com/mirego/trikot/datasources/flow/generic/FileSystemDataSource.kt [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+import okio.Path
+import okio.Path.Companion.toPath
+
+class FileSystemDataSource<R : ExpiringFlowDataSourceRequest, T>(
+    private val json: Json,
+    private val dataSerializer: KSerializer<T>,
+    private val diskCachePath: String
+) : BaseExpiringExecutableFlowDataSource<R, T>() {
+
+    private val fileManager = NativeFileSystem.fileSystem
```
Comment (npresseault) [inline trikot-datasources/datasources-flow/src/commonMain/kotlin/com/mirego/trikot/datasources/flow/generic/FileSystemDataSource.kt:22]: Could we allow the fileSystem to be passed in the construction and default to NativeFileSystem.fileSystem ?

**Review by npresseault — APPROVED** [review]

**Comment by npresseault** [comment]
Tests would've been nice too!

**Review by marcantoinefortier — APPROVED** [review]

**Pull request merged**